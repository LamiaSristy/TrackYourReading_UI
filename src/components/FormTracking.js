/* eslint-disable camelcase */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { updateDay } from '../actions/trackings';
import '../stylesheets/FormTracking.css';

class FormDay extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      date: '2021-03-10',
      page_no: 105,
    };
  }

  componentDidMount = () => {
    const { actionToPerform, buttonId, trackings } = this.props;
    if (actionToPerform === 'Save Changes') {
      const track = trackings.filter(x => x.id.toString() === buttonId);
      this.setState({
        date: track[0].date,
        page_no: track[0].page_no,
      });
    }
  }

  handleChangeDate = e => {
    this.setState({
      date: e.target.value,
    });
  }

  handleChangePageNo = e => {
    this.setState({
      page_no: e.target.value,
    });
  }

  handleEdit = async (id, book_id) => {
    const {
      date, page_no,
    } = this.state;
    const {
      user, updateDay, changeEditForm,
    } = this.props;
    const data = {
      id,
      user_id: user.user.id,
      book_id,
      date,
      page_no,
    };

    await updateDay(data);
    changeEditForm();
  }

  handleSubmit(date, page_no) {
    const { addTracking, changeAddForm, user } = this.props;
    const user_id = user.user.id;
    addTracking(date, page_no, user_id);
    changeAddForm();
  }

  render() {
    const {
      date, page_no,
    } = this.state;
    const {
      actionToPerform, trackings, buttonId, changeAddForm, changeEditForm,
    } = this.props;
    const track = trackings.filter(x => x.id.toString() === buttonId);

    return (
      <div>
        <p>
          {actionToPerform}
          {' '}
          Tracking for Books
        </p>
        <form className="day">

          <div className="form-div">
            <div className="date-track">
              <div className="date-div">
                <label htmlFor="date">
                  Date:
                  <input
                    id="date"
                    type="date"
                    name="date"
                    defaultValue={buttonId === '0' ? date : track[0].date.slice(0, 10)}
                    onChange={this.handleChangeDate}
                  />
                </label>

              </div>
              <div className="page_no-div">
                <label htmlFor="pageno">
                  No. of Page:
                  <input
                    id="pageno"
                    type="number"
                    name="page_no"
                    defaultValue={buttonId === '0' ? page_no : track[0].page_no}
                    onChange={this.handleChangePageNo}
                  />
                </label>
              </div>
            </div>
            <div className="buttons-form day-buttons">
              {actionToPerform === 'Add' && <button type="button" onClick={() => this.handleSubmit(date, page_no)}>{actionToPerform}</button>}
              {actionToPerform === 'Save Changes' && <button type="button" onClick={() => this.handleEdit(track[0].id, track[0].book_id)}>Save</button>}
              {actionToPerform === 'Add' && <button type="button" onClick={changeAddForm}>Cancel</button>}
              {actionToPerform === 'Save Changes' && <button type="button" onClick={changeEditForm}>Cancel</button>}
            </div>
          </div>
        </form>
      </div>
    );
  }
}

FormDay.propTypes = {
  addTracking: PropTypes.func,
  changeEditForm: PropTypes.func,
  changeAddForm: PropTypes.func,
  updateDay: PropTypes.func,
  buttonId: PropTypes.string,
  actionToPerform: PropTypes.string,
  trackings: PropTypes.instanceOf(Array),
  user: PropTypes.shape({
    user: PropTypes.shape({
      id: PropTypes.number,
    }),
  }),
};

FormDay.defaultProps = {
  changeAddForm: () => {},
  addTracking: () => {},
  updateDay: () => {},
  changeEditForm: () => {},
  actionToPerform: '',
  trackings: [],
  buttonId: '0',
  user: {},

};

const mapStateToProps = state => ({
  user: state.user,
  trackings: state.tracking,
});
const mapDispatchToProps = dispatch => ({
  updateDay: data => dispatch(updateDay(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(FormDay);
