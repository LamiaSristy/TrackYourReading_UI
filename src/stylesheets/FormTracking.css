@import "https://fonts.googleapis.com/css2?family=Roboto&display=swap";

.day {
  font-family: 'Roboto', sans-serif;
  display: flex;
  justify-content: flex-end;
  border: 1px solid gray;
  background-color: white;
  margin: 0 15px 12px 15px;
  border-radius: 11px;
  flex-direction: column;
  font-size: 3.5vw;
}

input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.day input {
  border-radius: 5px;
  border: 1px solid gray;
  font-size: 3.2vw;
}

.date-track {
  display: flex;
  margin: 8px 0 8px 7px;
  justify-content: center;
}

.date-div,
.page_no-div {
  display: flex;
  align-items: center;
}

.date-div {
  margin-right: 8px;
}

.date-div input,
.page_no-div input {
  width: 32vw;
  font-size: 3.2vw;
  font-family: 'Roboto', sans-serif;
  padding: 6px;
}

.page_no-div input {
  width: 9vw;
}

.day-buttons {
  justify-content: center;
}

.day-buttons button {
  font-size: 3.5vw;
  width: 23vw;
  cursor: pointer;
}

input[type="date"]::-webkit-calendar-picker-indicator {
  display: block;
  cursor: pointer;
}
