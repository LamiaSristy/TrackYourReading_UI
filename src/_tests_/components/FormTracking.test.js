import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import FormTracking from '../../components/FormTracking';

const mockStore = configureMockStore();
const store = mockStore({});
const setup = () => shallow(
  <Provider store={store}>
    <FormTracking />
  </Provider>,
);

describe('FormTracking Component', () => {
  let component;
  beforeEach(() => {
    component = setup();
  });

  test('renders the day the page', () => {
    const page = component.find('.day');
    expect(page).toMatchSnapshot();
  });

  test('renders the text of the date-track class', () => {
    const text = component.find('.date-track');
    expect(text).toMatchSnapshot();
  });

  test('renders the text of the date-div class', () => {
    const text = component.find('.date-div');
    expect(text).toMatchSnapshot();
  });

  test('renders the text of the page_no-div class', () => {
    const text = component.find('.page_no-div');
    expect(text).toMatchSnapshot();
  });

  test('renders the text of the day-buttons class', () => {
    const text = component.find('.day-buttons');
    expect(text).toMatchSnapshot();
  });
});
