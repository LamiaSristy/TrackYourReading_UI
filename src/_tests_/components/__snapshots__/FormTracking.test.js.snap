// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTracking Component renders the day the page 1`] = `ShallowWrapper {}`;

exports[`FormTracking Component renders the text of the buttons-form class 1`] = `ShallowWrapper {}`;

exports[`FormTracking Component renders the text of the date-div class 1`] = `ShallowWrapper {}`;

exports[`FormTracking Component renders the text of the date-div class 2`] = `ShallowWrapper {}`;

exports[`FormTracking Component renders the text of the date-div class 3`] = `ShallowWrapper {}`;

exports[`FormTracking Component renders the text of the date-track class 1`] = `ShallowWrapper {}`;

exports[`FormTracking Component renders the text of the day-buttons class 1`] = `ShallowWrapper {}`;

exports[`FormTracking Component renders the text of the one-parameter class 1`] = `ShallowWrapper {}`;

exports[`FormTracking Component renders the text of the page_no-div class 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the form the page 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the text of the buttons-form class 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the text of the one-parameter class 1`] = `ShallowWrapper {}`;
